### DISCUSSION TITLE
Matching `^` and `$` with `Regex::captures_at`

### DISCUSSION THREAD

**Discussion by ClementNerma** [rust-lang/regex#1125] [Q&A] [answered]
Hi there!

I'm currently building a tiny TextMate-like syntax highlighter using regexes at its core.

One thing I need to do is to match the _beginning of a substring_. For instance:

```rust
let regex = Regex::new("^@direct");
let input = "call(@direct cmd)";

regex.captures_at(input, 5); // Must match
```

Unfortunately, this doesn't work due to `captures_at` being aware of the input context, which means `^`  will only match if we're actually at the beginning of the _whole input_, not just the part we are matching against.

What workaround could be used for this use case? I need something that will allow the regex to match in the exact same conditions (full input, provided position).

**EDIT:** To be clearer, I need a way to specify in the regex that I want to match the beginning or end of the _portion of string_ it is matched against, not the _full string_.

Thanks in advance for your help!

**Comment by BurntSushi** [comment, marked as answer]
I think the direct and most straight-forward answer to your _specific_ question is `regex.captures(&input[5..])`.

Otherwise, if you want to start a search at an arbitrary position while taking look-behind assertions into account but want to limit your matches to those that start at the same position that you started your search, then you'll have to drop down to `regex-automata` and use its [anchored option](https://docs.rs/regex-automata/latest/regex_automata/struct.Input.html#method.anchored). The example there uses the `PikeVM`, but it will also work with the [meta regex engine](https://docs.rs/regex-automata/latest/regex_automata/meta/index.html) which is probably what you want to use. (The `regex` crate is itself just a wrapper around the meta regex engine.)
